06_puppy/nicholascross/main.go

```diff
@@ -0,0 +1,87 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"sync"
+)
+
+type Puppy struct {
+	ID     uint64
+	Breed  string
+	Colour string
+	Value  float64
+}
+
+type Storer interface {
+	CreatePuppy(p Puppy)
+	RetrievePuppy(uint64) *Puppy
+	UpdatePuppy(uint64, Puppy)
+	DeletePuppy(uint64)
+}
+
+type MapStore map[uint64]Puppy
+
+var out io.Writer = os.Stdout
+
+func main() {
+	p := Puppy{1, "Jack Russel Terrier", "white and brown", 550.0}
+	fmt.Fprintf(out, "%d - %s [%s]: %f", p.ID, p.Breed, p.Colour, p.Value)
+}
+
+func (ms MapStore) CreatePuppy(p Puppy) {
+	if _, exists := ms[p.ID]; !exists {
+		ms[p.ID] = p
+	}
+}
+
+func (ms MapStore) RetrievePuppy(id uint64) *Puppy {
+	if _, exists := ms[id]; exists {
+		pup := ms[id]
+		return &pup
+	}
+	return nil
+}
+
+func (ms MapStore) UpdatePuppy(id uint64, p Puppy) {
+	if _, exists := ms[id]; exists {
+		p.ID = id
+		ms[p.ID] = p
+	}
+}
+
+func (ms MapStore) DeletePuppy(id uint64) {
+	if _, exists := ms[id]; exists {
```

Comment:
There is no need for this check. It is safe to delete an element that does not exist: https://golang.org/pkg/builtin/#delete